app/services/listing_service/store/shapes.rb

```diff
@@ -0,0 +1,147 @@
+  ListingUnitModel = ::ListingUnit
+
+  NewShape = EntityUtils.define_builder(
+    [:community_id, :fixnum, :mandatory],
+    [:price_enabled, :bool, :mandatory],
+    [:name_tr_key, :string, :mandatory],
+    [:action_button_tr_key, :string, :mandatory],
+    [:transaction_process_id, :fixnum, :mandatory],
+    [:translations, :array, :optional], # TODO Only temporary
+    [:units, :array, default: []], # Mandatory only if price_enabled
+  )
+
+  Shape = EntityUtils.define_builder(
+    # TODO Currently we don't have Shape model, i.e. we don't have Shape id
+    # [:id, :fixnum, :mandatory]
+    [:transaction_type_id, :fixnum, :optional], # TODO Only temporary
+    [:community_id, :fixnum, :mandatory],
+    [:price_enabled, :to_bool, :mandatory], # to_bool, because there are NULL values in db
+    [:name_tr_key, :string, :mandatory],
+    [:action_button_tr_key, :string, :mandatory],
+    [:transaction_process_id, :fixnum, :mandatory],
+    [:translations, :array, :optional], # TODO Only temporary
+    [:units, :array, :mandatory]
+  )
+
+  Unit = EntityUtils.define_builder(
+    [:type, :to_symbol, one_of: [:piece, :hour, :day, :night, :week, :month, :custom]],
+    [:translation_key, :optional] # TODO Validate or transform to TranslationKey
+    )
+
+  module_function
+
+  # TODO Remove transaction_type_id
+  def get(community_id:, transaction_type_id: nil, listing_shape_id: nil)
+    if transaction_type_id
+      model = TransactionTypeModel.where(community_id: community_id, id: transaction_type_id).first
+      from_transaction_type_model(model)
+    elsif listing_shape_id
+      raise NotImplementedError.new("Can not find listing shape by listing_shape_id, yet. Specify transaction_type_id instead.")
+    else
+      rase ArgumentError.new("Can not find listing shape without id.")
+    end
+  end
+
+  def create(community_id:, opts:)
+    shape = NewShape.call(opts.merge(community_id: community_id))
+
+    valid_to_create!(shape)
+
+    units = shape[:units].map { |unit| Unit.call(unit) }
+    translations = opts[:translations] # Skip data type and validation, because this is temporary
+
+    # TODO We should be able to create transaction_type without community
+    community = Community.find(shape[:community_id])
+
+    create_tt_opts = to_tt_model_attributes(shape).except(:units, :translations)
+    tt_model = community.transaction_types.build(create_tt_opts)
+
+    raise NotImplementedError.new("For backward compatibility reasons saving multiple units is not yet supported") if units.length > 1
+
+    units.each { |unit|
+      tt_model.listing_units.build(to_unit_model_attributes(unit))
+    }
+    translations.each { |tr| tt_model.translations.build(tr) }
+
+    tt_model.save!
+
+    from_transaction_type_model(tt_model)
+  end
+
+  # private
+
+  def valid_to_create!(shape)
+    valid = (shape[:price_enabled] == true && !shape[:units].empty?) || (shape[:price_enabled] == false && shape[:units].empty?)
```

Comment:
Good point.

Now when I think this, at least the first condition is wrong. If there is a listing e.g. "House in the middle of Helsinki", the listing will have price but NOT units. Even though the price is per piece, I think it doesn't make sense to have the per piece unit, because there's only one house to sell.

About the second condition, and your example, that's also a good question. Probably need to ask Juho, or maybe I just remove this whole validation method and rely that the caller calls the API with correct params.